src/bindings/python/flux/future.py

```diff
@@ -185,6 +187,55 @@ def get(self):
     def incref(self):
         self.pimpl.flux_future_incref()
 
+    # asyncio functions
+
+    def __await__(self):
+        """
+        Await to work with asyncio.
+        """
+        if not hasattr(self, "_loop"):  # pylint: disable=attribute-defined-outside-init
+            import flux.asyncio  # pylint: disable=import-outside-toplevel
+            self._loop = flux.asyncio.loop  # pylint: disable=attribute-defined-outside-init
+
+        if not self.is_ready():
+            self._asyncio_future_blocking = True  # pylint: disable=attribute-defined-outside-init
+
+            # Yielding here tells the task to wait for completion.
+            yield self
+
+        if not self.is_ready():
+            raise RuntimeError("await wasn't used with future")
+        return self.get()
+
+    def add_done_callback(self, func, *, context=None):
```

Comment:
I don't think you should add methods to the existing `Future` class, because these methods might be mistakenly called by users who aren't using the asyncio interface, and that would be pretty confusing, especially given the kinds of errors that would result. I think it would be better to either wrap the `Future` in another class, or subclass it.

Reply:
I thought the goal was to use the existing Flux future and just tweak slightly to work with asyncio - to be clear this absolutely won't work without these methods. Pinging @trws for comment because we made this decision together.

Reply:
If we made another Future class (that adds the methods) it wouldn't plug nicely into submit_async (unless we did some weird fu)

Reply:
Yeah I take back my subclass idea, that doesn't make sense.

I guess what I was thinking is that instead of what you currently have:

```python

async def submit(fluxjob):
    uid = await flux.job.submit_async(handle, fluxjob)
    return uid
```

you could do something like


```python

class Wrapper:
    [stuff]

    def __init__(self, future):
       self.future = future

    def __await__(self)
        [stuff with self.future]


async def submit(fluxjob):
    loop = asyncio.get_running_loop()
    flux_handle = loop.flux_handle
    w = Wrapper(flux.job.submit_async(handle, fluxjob), loop)
    return await w
```

Somewhat related: when you need the loop, instead of trying to reach out to `flux.asyncio.loop`, you could just call `asyncio.get_running_loop()` which I think is more portable.

Also, if you make the loop's `flux.Flux()` handle a public attribute, you can enforce that `submit` and the loop` always have the same flux handle. 

I think this should actually work pretty nicely, but let me know if I'm missing something.

Reply:
oh nice! Let me try that. Just to clarify - this should be like this right?
```diff
- flux_handle = loop.flux_handle
+  handle = loop.flux_handle
```
or I can just reference `loop.flux_handle` directly in the call.  I'll test the above now - I'm not sure how well it's going to wrok moving the await there but we'll find out!

Reply:
Also where does the loop come in? it looks like you provided it as a second arg, but wrapper only takes 1 argument (the fluxjob).

Reply:
Yeah I just threw it together in a hurry, sorry for all the silly errors. I hadn't decided whether it would be better to pass the loop around all the time or instead to call `get_running_loop` whenever you needed it

Reply:
okay it didn't like our wrapper:

> RuntimeError: Task got bad yield: <flux.job.submit.SubmitFuture object at 0x7fccaaa67b50>

Reply:
Do you want me to push this somewhere for you to mess around with @jameshcorbett ?

Reply:
You could probably just as easily start with the current version. I do want to restore what I had that was working (it was really hard to figure out).

Reply:
okay - I updated to use asyncio.get_running_loop() - a huge improvement over the circular import! Nice!

Reply:
Yeah restore whatever you had, I will play around and perhaps propose a patch later, but I need to look at other stuff now.

Reply:
okay sounds good! Thanks for your help today we made really good progress! :partying_face:

Reply:
This would work, but I strongly prefer having our futures directly support the standard python futures interface.  Note that they generally are not asyncio specific.  This one currently is because of the invocation of loop, but we should be able to factor it into an use of either `.then` if the future is not ready or adding a `TimedWatcher` to the associated handle.  Actually, that's a question, @garlick what happens if you call `flux_future_then` or similar on a future that's already ready?

Either way, if we make it so that the future provides the "normal" python futures interface on top of flux features, in addition to our bespoke interface from before this stuff settled, I think we'd be better off overall *and* we get asyncio support without extra wrapper calls, just directly `await flux.job.submit()` or `loop.run_until_complete(asyncio.wait(loop.create_task(flux.job.submit())))`.

Reply:
> if we make it so that the future provides the "normal" python futures interface on top of flux features, in addition to our bespoke interface from before this stuff settled, I think we'd be better off overall and we get asyncio support without extra wrapper calls, just directly await flux.job.submit() or loop.run_until_complete(asyncio.wait(loop.create_task(flux.job.submit()))).

That makes a lot of sense actually, hadn't thought about it that way.